cfg/CFG.cc

```diff
@@ -22,97 +32,161 @@ BasicBlock *CFG::freshBlock(int outerLoops, int rubyBlockId) {
-                blockReadsAndWrites.insert(v->send);
+                blockReads.insert(v->send.id());
             }
 
-            auto fnd = blockReads.find(bind.bind.variable);
-            if (fnd == blockReads.end()) {
-                blockDead.insert(bind.bind.variable);
+            if (!blockReads.contains(bind.bind.variable.id())) {
+                blockDead.insert(bind.bind.variable.id());
             }
         }
         ENFORCE(bb->bexit.cond.variable.exists());
-        if (bb->bexit.cond.variable != core::LocalVariable::unconditional()) {
-            blockReads.insert(bb->bexit.cond.variable);
-            blockReadsAndWrites.insert(bb->bexit.cond.variable);
+        if (bb->bexit.cond.variable != LocalRef::unconditional()) {
+            blockReads.insert(bb->bexit.cond.variable.id());
         }
+
+        // Convert sets to sorted vectors.
+        auto &blockReadsVector = target.reads[bb->id];
+        blockReadsVector.reserve(blockReads.size());
+        blockReadsVector.insert(blockReadsVector.end(), blockReads.begin(), blockReads.end());
+        fast_sort(blockReadsVector);
+
+        auto &blockWriteVector = target.writes[bb->id];
+        blockWriteVector.reserve(blockWrites.size());
+        blockWriteVector.insert(blockWriteVector.end(), blockWrites.begin(), blockWrites.end());
+        fast_sort(blockWriteVector);
+
+        auto &blockDeadVector = target.dead[bb->id];
+        blockDeadVector.reserve(blockDead.size());
+        blockDeadVector.insert(blockDeadVector.end(), blockDead.begin(), blockDead.end());
+        fast_sort(blockDeadVector);
     }
-    UnorderedMap<core::LocalVariable, pair<int, int>> usageCounts;
+
+    vector<pair<int, int>> usageCounts(maxVariableId);
 
     {
         Timer timeit(ctx.state.tracer(), "privates1");
 
-        for (auto blockId = 0; blockId < maxBasicBlockId; blockId++) {
-            for (auto &el : readsAndWrites[blockId]) {
-                usageCounts.try_emplace(el, make_pair(0, blockId)).first->second.first += 1;
+        for (u4 blockId = 0; blockId < maxBasicBlockId; blockId++) {
+            const auto &blockReads = target.reads[blockId];
+            const auto &blockWrites = target.writes[blockId];
+            vector<int> blockReadsAndWrites;
+            blockReadsAndWrites.reserve(max(blockReads.size(), blockWrites.size()));
+            set_union(blockReads.begin(), blockReads.end(), blockWrites.begin(), blockWrites.end(),
+                      back_inserter(blockReadsAndWrites));
+            for (auto local : blockReadsAndWrites) {
+                if (usageCounts[local].first == 0) {
+                    usageCounts[local].second = blockId;
+                }
+                usageCounts[local].first += 1;
             }
         }
     }
     {
         Timer timeit(ctx.state.tracer(), "privates2");
-        for (const auto &[local, usages] : usageCounts) {
+        auto local = 0;
+        vector<vector<int>> writesToRemove(maxBasicBlockId);
```

Comment:
Individual item removal from a sorted list is `O(log n)` (if you use binary search) or `O(n)` (if you iterate in-order). But if you're removing a list of sorted items from a sorted list, it's just `O(n)` because you can remove all of them in one pass.